#506 BkTimeableTest fails intermittently

raphaelln opened the pull request:
The principal problem was in the RqMultipartTest because the requests are creating temporary files greater than 97mb, those files are now deleted after the junit test.

But, some files with 1KB stills remains in computer, I believe we need to fix the issue #576.

**- Refactoring notes**
 1) Some tests was creating files not using TemporaryFolder.
 2) RqMultipartTest was change to delete the temporary files after the execution of the junit. In order to that, I need the close the body part of the multipart request.

With these changes, after the execution of the all junits, the temporay files with 97mb are deleted after the junit ends, eliminating the instability of the builds. 

**This PR is related with issue #506**

**Code changes**

- RqMultipartTest - Lines 106 - 117
  Without that changes the code result in two more files created and not deleted after the junit test ends. So we need to close the body of this parts in order to delete them after we used them (#576).

- BkTimeableTest: add a Custom exit condition, to latch the current thread to continue the process of the junit.

davvd commented:
@raphaelln Thanks, let me find someone who can review this pull request

davvd commented:
@mkordas please review, thanks

mkordas commented:
@raphaelln I'm on it

mkordas commented:
@raphaelln please refer to issue from PR description

mkordas commented on src/test/java/org/takes/facets/fork/FkHitRefreshTest.java:
```diff
@@ -39,12 +40,19 @@
 
     /**
+     * Temp directory.
+     * @checkstyle VisibilityModifierCheck (5 lines)
```
@raphaelln do we need this suppression?

mkordas commented on src/test/java/org/takes/facets/fork/FkHitRefreshTest.java:
```diff
@@ -39,12 +40,19 @@
  * Test case for {@link FkHitRefresh}.
  * @author Yegor Bugayenko (yegor@teamed.io)
- * @version $Id$
+ * @version $Id: e60e0217ac71dbde8b13fc678e4770cbd2f9982c $
```
@raphaelln please revert this change

mkordas commented on src/test/java/org/takes/facets/fork/FkHitRefreshTest.java:
```diff
@@ -81,11 +88,11 @@ public void run() {
             new FkHitRefresh(
                 dir, "", new TkEmpty()
-            ).route(new RqFake()).has(),
+                ).route(new RqFake()).has(),
```
@raphaelln why you modified this line?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -52,7 +54,7 @@
  * Test case for {@link RqMultipart.Base}.
  * @author Yegor Bugayenko (yegor@teamed.io)
- * @version $Id$
+ * @version $Id: 37f595f280dc032586806cdefa9afea1f279d9e0 $
```
@raphaelln please revert

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -95,10 +106,22 @@ public void satisfiesEqualsContract() throws IOException {
+            req.body().close();
+            reqbaseone.part("t-1")
+            .iterator()
+            .next().body().close();
```
@raphaelln something is wrong with indentation here, please fill one line with as much as possible

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -206,23 +229,28 @@ public void parsesHttpBody() throws IOException {
-                Matchers.endsWith("CA 94085")
-            )
-        );
+                    ).header(RqMultipartTest.DISPOSITION),
```
@raphaelln wrong indentation

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -279,12 +311,16 @@ public void returnsCorrectNamesSet() throws IOException {
+                multi.names(),
+                Matchers.<Iterable<String>>equalTo(
+                    new HashSet<String>(Arrays.asList("address", "data"))
+                    )
```
@raphaelln please move 4 spaces back

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -494,22 +545,30 @@ public void producesPartsWithContentLength() throws IOException {
+                new HmRqHeader(
+                    "content-length",
+                    "102"
+                    )
```
@raphaelln 4 spaces back

mkordas commented:
@raphaelln please provide more description to PR, you did quite a lot of changes, I need to understand why

mkordas commented on src/test/java/org/takes/facets/fork/FkHitRefreshTest.java:
```diff
@@ -53,11 +61,10 @@
             new RqFake(), "X-Takes-HitRefresh: yes"
-        );
-        final File dir = Files.createTempDir();
+            );
```
@raphaelln fix indentation

mkordas commented on src/test/java/org/takes/facets/fork/FkHitRefreshTest.java:
```diff
@@ -65,9 +72,9 @@ public void run() {
             },
             new TkEmpty()
-        );
+            );
```
@raphaelln revert that change

mkordas commented on src/test/java/org/takes/tk/TkFilesTest.java:
```diff
@@ -43,21 +43,27 @@
 
     /**
+     * Temp directory.
+     * @checkstyle VisibilityModifierCheck (5 lines)
```
@raphaelln remove suppression

mkordas commented on src/test/java/org/takes/tk/TkFilesTest.java:
```diff
@@ -43,21 +43,27 @@
+                new TkFiles(this.temp.getRoot()).act(
                     new RqFake(
                         "GET", "/a.txt?hash=a1b2c3", ""
+                        )
```
@raphaelln this is wrong indentation, please read http://www.yegor256.com/2014/10/23/paired-brackets-notation.html

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -494,22 +545,30 @@ public void producesPartsWithContentLength() throws IOException {
+        } finally {
+            multipart.body().close();
+            multipart.part(RqMultipartTest.PART_T2).iterator().next()
+            .body().close();
```
@raphaelln add 4 spaces indentation here

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -494,22 +545,30 @@ public void producesPartsWithContentLength() throws IOException {
-                    "Content-Disposition: form-data; name=\"t2\"",
+                    new StringBuilder()
+                    .append("Content-Disposition: form-data; name=\"")
+                    .append(RqMultipartTest.PART_T2).append("\""),
```
@raphaelln why this change is needed? please explain in PR description

raphaelln replied:
after the changes the word "t2" appeared twice in the code, so I need to create a constant, and concat with Strings are not possible, so I used the StringBuilder to do that.

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -494,22 +545,30 @@ public void producesPartsWithContentLength() throws IOException {
                 Joiner.on("\r\n").join(
                     "--AaB01x",
-                    "Content-Disposition: form-data; name=\"t2\"",
+                    new StringBuilder()
```
@raphaelln why do we need `StringBuilder` here?

raphaelln replied:
after the changes the word "t2" appeared twice in the code, so I need to create a constant, and concat with Strings are not possible, so I used the StringBuilder to do that.

mkordas replied:
@raphaelln please add the name at the beginning of message

mkordas replied:
@raphaelln please read http://www.yegor256.com/2014/06/19/avoid-string-concatenation.html - `StringBuilder` is not a proper replacement

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -459,23 +506,27 @@ public void notDistortContent() throws IOException {
+                MatcherAssert.assertThat(
+                    String.format("byte %d not matched", idx),
+                    stream.read(),
+                    Matchers.equalTo(idx % byt)
```
@raphaelln minor: usually we use `Matchers.is`

raphaelln replied:
I didn't change from the previous revision released version.

mkordas replied:
@raphaelln please add '@mkordas' above

mkordas replied:
@raphaelln I see nevermind then

mkordas replied:
@raphaelln please add '@mkordas' above

mkordas replied:
@raphaelln please update the following comment with addressee:
> I didn't change from the previous revision released version.

mkordas replied:
@raphaelln I still don't see your comment updated

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -459,23 +506,27 @@ public void notDistortContent() throws IOException {
-            Matchers.equalTo(length)
-        );
-        for (int idx = 0; idx < length; ++idx) {
+            ).single("test1").body();
```
@raphaelln again, move back

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -403,23 +445,28 @@ public void handlesRequestInTime() throws IOException {
+            );
+        final RqMultipart.Smart smartreq = new RqMultipart.Smart(
+            new RqMultipart.Base(req)
+            );
```
@raphaelln please pay attention to indentation here as well

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -311,12 +347,18 @@ public void returnsCorrectPartLength() throws IOException {
-                new RqMultipart.Base(req)
-            ).single("x-1").body().available(),
-            Matchers.equalTo(length)
+        final RqMultipart.Smart regsmart =  new RqMultipart.Smart(
```
@raphaelln double space, please fix

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -70,6 +75,12 @@
     private static final String DISPOSITION = "Content-Disposition";
+    /**
+     * Temp directory.
+     * @checkstyle VisibilityModifierCheck (5 lines)
```
@raphaelln remove suppression

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -70,6 +75,12 @@
      * Content disposition.
      */
     private static final String DISPOSITION = "Content-Disposition";
+    /**
```
@raphaelln add newline between fields

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -61,7 +63,10 @@
+    /**
+     * T2 temporary part name.
+     */
+    private static final String PART_T2 = "t2";
```
@raphaelln newline here as well

mkordas commented:
@raphaelln please see several comments from my side above

raphaelln commented:
@mkordas thanks for your review and sorry for the checkstyle errors.

I corrected the mistakes, changed the PR description and justified some changes.

Thanks.

mkordas commented:
@raphaelln thanks, let me review again

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -63,6 +65,10 @@
     /**
+     * T2 temporary part name.
+     */
+    private static final String PART_T2 = "t2";
```
@raphaelln please convert it into local variable instead. It should be limited to scope of just one test

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -95,10 +107,18 @@ public void satisfiesEqualsContract() throws IOException {
+                Matchers.equalTo(reqbasetwo)
+            );
+        } finally {
+            req.body().close();
```
@raphaelln why these changes are needed? Can you point me to specific sentence in PR description?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -95,10 +107,18 @@ public void satisfiesEqualsContract() throws IOException {
-            new RqMultipart.Base(req),
-            Matchers.equalTo(new RqMultipart.Base(req))
-        );
+        final RqMultipart.Base reqbaseone = new RqMultipart.Base(req);
```
@raphaelln please avoid compound names: http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -206,23 +226,28 @@ public void parsesHttpBody() throws IOException {
-        );
+            );
+        } finally {
+            multi.body().close();
```
@raphaelln do we really need to close all of that requests?

raphaelln replied:
You are right, just the part should be closed.

mkordas replied:
@raphaelln please do not forget about nickname to who you are talking to

mkordas replied:
@raphaelln ping here

mkordas commented:
@raphaelln please see additional comments above

mkordas commented:
@raphaelln ping

raphaelln commented:
@mkordas please could you review again ?

mkordas commented:
@raphaelln sure

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -95,10 +103,18 @@ public void satisfiesEqualsContract() throws IOException {
+        try {
+            MatcherAssert.assertThat(
+                one,
+                Matchers.equalTo(two)
```
@raphaelln please join these lines now

mkordas commented:
@raphaelln do I undersand correctly, are you fixing #576 here?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -206,23 +222,27 @@ public void parsesHttpBody() throws IOException {
-        );
+            );
+        } finally {
+            multi.part("t4").iterator().next().body().close();
```
@raphaelln we have duplicated `"t4"` here, why Checkstyle is not complaining?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -249,10 +269,14 @@ public void returnsEmptyIteratorOnInvalidPartRequest() throws IOException {
+                multi.part("fake").iterator().hasNext(),
+                Matchers.is(false)
+            );
+        } finally {
```
@raphaelln I'm not convinced about these finally blocks... Can't we just invoke `multi.body().close();` without finally?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -311,12 +339,18 @@ public void returnsCorrectPartLength() throws IOException {
+            );
+        } finally {
+            req.body().close();
+            regsmart.part("x-1").iterator().next().body().close();
```
@raphaelln `"x-1"` duplicated

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -378,8 +412,8 @@ public void exec(final URI home) throws IOException {
         final int length = 100000000;
-        final File temp = File.createTempFile("handlesRequestInTime", ".tmp");
-        final BufferedWriter bwr = new BufferedWriter(new FileWriter(temp));
+        final File tempfile = this.temp.newFile("handlesRequestInTime.tmp");
```
@raphaelln compound name, let's leave just `temp`

raphaelln replied:
checkstyle complains about omission of "this" word.

mkordas replied:
@raphaelln please update comment above

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -403,23 +437,28 @@ public void handlesRequestInTime() throws IOException {
-            System.currentTimeMillis() - start,
-            //@checkstyle MagicNumberCheck (1 line)
-            Matchers.lessThan(3000L)
+        final RqMultipart.Smart smartreq = new RqMultipart.Smart(
```
@raphaelln compound name, let's use just `smart`

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -428,8 +467,8 @@ public void handlesRequestInTime() throws IOException {
         final int length = 1000000;
-        final File temp = File.createTempFile("notDistortContent", ".tmp");
-        final BufferedWriter bwr = new BufferedWriter(new FileWriter(temp));
+        final File tempfile = this.temp.newFile("notDistortContent.tmp");
```
@raphaelln just `temp` should be fine

raphaelln replied:
checkstyle complains about omission of "this" word.

mkordas replied:
@raphaelln please update comment here as well

mkordas replied:
@raphaelln it's not about removing `this`, please read http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

raphaelln commented:
> @raphaelln do I understand correctly, are you fixing #576 here?

No, I am not fixing it, this is just a note why we have to close the body parts. In order to fix the issue #506 I did theses changes, because when you run the unit tests a lot of times, the disk space of could be compromised.  The disk space are affecting the other tests like the test refered in the issue (#506).

raphaelln commented:
@mkordas please could you review again ?
I just did a change in BkTimeableTest, put an custom exit to latch the current thread to process the file in sequence (also commented in the PR description)

Thanks.

mkordas commented:
@raphaelln thanks, but aren't we putting too much in this PR? Shouldn't this latch be separate issue?

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -95,10 +112,15 @@ public void satisfiesEqualsContract() throws IOException {
+        final RqMultipart.Base two = new RqMultipart.Base(req);
+        try {
+            MatcherAssert.assertThat(one, Matchers.equalTo(two));
+        } finally {
```
@raphaelln same queston here - do we really need to use finally here? This just ensures closing being done on assertion failure, but on assertion failure test is failed anyway so we don't care that much it think.

The only exception would be if after assertion failure some files were stored on disk permanently.

raphaelln replied:
I believe the code inside finally needs to be executed, because without the close of the body of all the parts, the files will be on disk permanently.

mkordas replied:
@raphaelln ok, thanks

mkordas replied:
@raphaelln please update your previous message with addressee at the beginning

mkordas replied:
@raphaelln ping, please update the comment above

mkordas replied:
@raphaelln one more reminder, please update the following comment with username at the beginning:
> I believe the code inside finally needs to be executed, because without the close of the body of all the parts, the files will be on disk permanently.

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -340,7 +383,11 @@ public Response act(final Request req) throws IOException {
-                "Content-Disposition: form-data; name=\"f-1\"", "",
+                String.format(
+                    RqMultipartTest.CONTENT,
+                    part
```
@raphaelln please join lines

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -378,12 +425,16 @@ public void exec(final URI home) throws IOException {
-                "Content-Disposition: form-data; name=\"test\"",
+                String.format(
+                    RqMultipartTest.CONTENT,
+                    part
```
@raphaelln join lines here too

mkordas commented:
@raphaelln please see a few minors and one question - why we need finally? Maybe there's good explanation for that.

raphaelln commented:
@mkordas  
> @raphaelln thanks, but aren't we putting too much in this PR? Shouldn't this latch be separate issue?

No, this is part of the issue (BkTimeableTest fails intermittently), this will prevent the test use the file and socket connection before the ready of the Req in **FtCLI**. 

**BkTimeableTest fails intermittently** because of two reasons:

1. After a lot of executions of the junit, the disk maybe be full and the another tests could be no executed, and there was 3 classes with this issue, one of them and the worst was the **RqMultipartTest**
2. BkTimeableTest is try to get the file and socket connection before the read of the Ftcli opens a socket and create the file.

In conclusion, all this changes is to prevent the errors described above.

raphaelln commented:
@mkordas could you please review again, and thanks for you time in review this issue, sorry about the mistakes.

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -494,22 +553,31 @@ public void producesPartsWithContentLength() throws IOException {
+                    String.format(
+                        RqMultipartTest.CONTENT,
+                        part
+                    ),
```
@raphaelln join lines here as well, please do it always when possible (it's not automatically enforced yet)

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -340,7 +383,8 @@ public Response act(final Request req) throws IOException {
                 "--AaB0zz",
-                "Content-Disposition: form-data; name=\"f-1\"", "",
+                String.format(RqMultipartTest.CONTENT, part),
+                "",
```
@raphaelln join lines here

mkordas replied:
@raphaelln ping, join lines here

mkordas replied:
@raphaelln again, join lines here, not everything is finished

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -428,12 +478,16 @@ public void handlesRequestInTime() throws IOException {
+                String.format(
+                    RqMultipartTest.CONTENT,
+                    part
+                ),
```
@raphaelln join lines

mkordas commented:
@raphaelln no worries, just a few minors more from me - we are almost done

raphaelln commented:
@mkordas could you review again ?

mkordas commented:
@raphaelln OK, I'm on it again

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -70,6 +72,20 @@
+    private static final String CONTENT = String.format(
+        "%s: %s",
+        RqMultipartTest.DISPOSITION,
+        "form-data; name=\"%s\""
```
@raphaelln `        "%s: %s", RqMultipartTest.DISPOSITION, "form-data; name=\"%s\""` fits one line

mkordas commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -294,10 +326,14 @@ public void returnsCorrectNamesSet() throws IOException {
-                "Content-Disposition: form-data; name=\"x-1\"",
+                String.format(
+                    RqMultipartTest.CONTENT,
+                    part
```
@raphaelln join lines

mkordas commented:
@raphaelln why your new commits have "unknown" author?

mkordas commented:
@raphaelln moreover, your first commits have 'rnascimento' author - is it still you?

mkordas commented:
@raphaelln please see remarks above

raphaelln commented:
> @raphaelln moreover, your first commits have 'rnascimento' author - is it still you?

Yes, I don't know why the commit was wrong. Anyway, I made  a new commit.

Could you review again.

mkordas commented:
@raphaelln 
> Yes, I don't know why the commit was wrong

You need to pay attention to email setting of your Git locally: https://help.github.com/articles/setting-your-email-in-git/

mkordas commented:
@raphaelln please update your comments with nicknames

mkordas commented:
@raphaelln please see 3 comments above

raphaelln commented:
@mkordas I correct the github local configurations, in my last commit my name is right.

Is there anything else to correct ?

mkordas commented:
@raphaelln yes, there are couple more things, please see my comments

mkordas commented:
@raphaelln comment https://github.com/yegor256/takes/pull/601#discussion_r52842778 still doesn't have nickname

raphaelln commented:
@mkordas Is there anything else to correct ? 

Thanks

mkordas commented:
@raphaelln yes, this one also misses addressee: https://github.com/yegor256/takes/pull/601#discussion_r52842781

raphaelln commented:
@mkordas 
> @raphaelln yes, this one also misses addressee: #601 (comment)

Done.

mkordas commented:
@raphaelln I've found yet another one: https://github.com/yegor256/takes/pull/601#discussion_r52841754 - just make sure all comments start with addressee name

raphaelln commented:
@mkordas I am really sorry about all this mistakes, I searched all my notes and comments, and I put the username on in it. Could you please review again and one more time thank you for your patient.

mkordas commented:
@raphaelln no worries, now it looks OK :)

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @yegor256 Please confirm this.

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/7081-184831559)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/7081-184831559) (took me 21min)

mkordas commented:
@rultor deploy

rultor commented:
> @rultor deploy

@mkordas Thanks for your request. @yegor256 Please confirm this.

davvd commented:
@elenavolokhova please, review this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

davvd commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7081-185839125)

rultor commented:
> @rultor please deploy

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/7081-185839125) (took me 9min)

elenavolokhova commented:
@davvd Looks good!

davvd commented:
> @davvd Looks good!

@elenavolokhova thank you

davvd commented:
@mkordas paied 10 mins to @elenavolokhova for QA review (payment ID is `78641922`)... done, I added **1 hour and 40 mins** in payment `56d1e65915431a2ea30001b6`, 138 hours and 41 mins spent to complete this... extra minutes for review comments (c=85)... +100 added to your rating, current score is: [+7428](http://www.netbout.com/b/36080?open=rating)
